librad/src/net/peer.rs

```diff
@@ -0,0 +1,195 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![allow(dead_code)]
+
+use std::fmt;
+
+use serde::{
+    de::{MapAccess, Visitor},
+    Deserialize,
+    Deserializer,
+    Serialize,
+    Serializer,
+};
+
+use crate::{
+    hash::Hash,
+    keys::device,
+    paths::Paths,
+    uri::{self, RadUrn},
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Rev {
+    Git(git2::Oid),
+}
+
+impl Serialize for Rev {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Git(oid) => {
+                let mut seq = vec![0];
+                seq.extend_from_slice(oid.as_bytes());
+                seq.serialize(serializer)
+            },
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Rev {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let x = Vec::deserialize(deserializer)?;
+        match x.split_first() {
+            Some((tag, oid)) => match tag {
+                0 => git2::Oid::from_bytes(oid)
+                    .map(Self::Git)
+                    .map_err(serde::de::Error::custom),
+                _ => Err(serde::de::Error::custom("Unknown tag")),
+            },
+
+            None => Err(serde::de::Error::custom("No data")),
+        }
+    }
+}
+
+impl Into<uri::Protocol> for Rev {
+    fn into(self) -> uri::Protocol {
```

Comment:
I think this would be better as a method on `Rev` that borrows instead of owns. It would avoid cloning the data inside just to map it to the protocol.

Reply:
Good catch. Source of idiomatic polizei is tell me it need programme like this deed82ec9bc243eae29042030690e3e39c0a0636

Reply:
:police_officer: